Source/DafnyCore/Resolver/PreType/PreType.cs

```diff
@@ -0,0 +1,326 @@
+//-----------------------------------------------------------------------------
+//
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+//
+//-----------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using Bpl = Microsoft.Boogie;
+
+namespace Microsoft.Dafny {
+  public abstract class PreType {
+    public PreType Normalize() {
+      var t = this;
+      while (t is PreTypeProxy proxy && proxy.PT != null) {
+        t = proxy.PT;
+      }
+      return t;
+    }
+
+    public static bool Same(PreType a, PreType b) {
+      a = a.Normalize();
+      b = b.Normalize();
+      if (a is DPreType ap && b is DPreType bp && ap.Decl == bp.Decl) {
+        Contract.Assert(ap.Arguments.Count == bp.Arguments.Count);
+        for (var i = 0; i < ap.Arguments.Count; i++) {
+          if (!Same(ap.Arguments[i], bp.Arguments[i])) {
+            return false;
+          }
+        }
+        return true;
+      }
+      return a == b;
+    }
+
+    public PreType UrAncestor(PreTypeResolver ptResolver) {
+      Contract.Requires(ptResolver != null);
+      var pt = this;
+      while (true) {
+        pt = pt.Normalize();
+        if (pt is DPreType preType && preType.Decl is NewtypeDecl newtypeDecl) {
+          // expand the newtype into its base type
+          var subst = PreTypeSubstMap(newtypeDecl.TypeArgs, preType.Arguments);
+          var basePreType = ptResolver.Type2PreType(newtypeDecl.BaseType);
+          pt = basePreType.Substitute(subst);
+        } else {
+          // nothing more we can do
+          return pt;
+        }
+      }
+    }
+
+    public DPreType AsCollectionType() {
```

Comment:
Given the naming, I expected the return type to be `CollectionType`, not `DPreType`.

Reply:
Good point. I'll at least rename it to `AsCollectionPreType`.